File: MyLearning/DsaPractice/Graph/DistinctIslands.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace DsaPractice.Graph
{
    /*
     * Problem: Number of Distinct Islands
     * Source: https://www.geeksforgeeks.org/problems/number-of-distinct-islands/1
     * 
     * Description:
     * Given a boolean 2D matrix 'grid' of size n x m, find the number of distinct islands.
     * An island is a group of connected 1s (connected horizontally or vertically).
     * Two islands are considered distinct if and only if one island is not equal to another
     * (no rotation or reflection is allowed).
     * 
     * Example:
     * Input:
     * grid[][] = { {1, 1, 0, 0, 0},
     *              {1, 1, 0, 0, 0},
     *              {0, 0, 0, 1, 1},
     *              {0, 0, 0, 1, 1} }
     * Output:
     * 1
     */
    internal class DistinctIslands
    {
        public int CountDistinctIslands(List<List<int>> grid)
        {
            int rows = grid.Count;
            int cols = grid[0].Count;

            HashSet<string> shapes = new();

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    StringBuilder path = new();

                    DFS(grid, row, col, rows, cols, path, "START");

                    if (path.Length > 0)
                        shapes.Add(path.ToString());
                }
            }

            return shapes.Count;
        }

        private void DFS(List<List<int>> grid, int row, int col, int rows, int cols, StringBuilder path, string direction)
        {
            if (row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] == 0)
                return;

            grid[row][col] = 0;

            // path.Append($"{row - startRow} : {col - startCol}"); // Another approach to do this

            path.Append(direction);

            DFS(grid, row - 1, col, rows, cols, path, "UP");
            DFS(grid, row + 1, col, rows, cols, path, "DOWN");
            DFS(grid, row, col - 1, rows, cols, path, "LEFT");
            DFS(grid, row, col + 1, rows, cols, path, "RIGHT");

            path.Append('B'); // Also include back track history
        }

        public void Test()
        {
            List<List<int>> grid =
            [[ 1, 1, 0, 0, 0],
            [ 1, 1, 0, 0, 0],
            [ 0, 0, 0, 1, 1],
            [ 0, 0, 0, 1, 1]];

            int distinctIslands = CountDistinctIslands(grid);

            Console.WriteLine(distinctIslands);
        }
    }
}

File: MyLearning/PythonPractice/Graph/DistinctIslands.py
from typing import List, Set, Tuple


def count_distinct_islands(grid: List[List[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    visited: Set[Tuple[int, int]] = set()
    shapes: Set[Tuple[Tuple[int, int], ...]] = set()

    def dfs(r: int, c: int, base_r: int, base_c: int, shape: List[Tuple[int, int]]) -> None:
        visited.add((r, c))
        shape.append((r - base_r, c - base_c))
        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = r + dr, c + dc
            if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == 1 and (nr, nc) not in visited:
                dfs(nr, nc, base_r, base_c, shape)

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == 1 and (r, c) not in visited:
                shape: List[Tuple[int, int]] = []
                dfs(r, c, r, c, shape)
                shapes.add(tuple(shape))
    return len(shapes)
